File: DID.Swift.SDK/ElastosDIDSDK/DID/Backend/IDChainRequest.swift
import Foundation

public class IDChainRequest: NSObject {
    public static let CURRENT_SPECIFICATION = "elastos/did/1.0"
    
    private static let HEADER = "header"
    private static let SPECIFICATION = "specification"
    private static let OPERATION = "operation"
    private static let PREVIOUS_TXID = "previousTxid"
    private static let PAYLOAD = "payload"
    private static let KEY_TYPE = TYPE
    private static let VERIFICATION_METHOD = "verificationMethod"
    private static let SIGNATURE = "signature"
    // private static let KEY_ID: String = VERIFICATION_METHOD
    
    // header
    public var specification: String = ""
    public var operation: IDChainRequest.Operation
    public var previousTxid: String = ""

    // payload
    public var did: DID?
    public var doc: DIDDocument?
    public var payload: String = ""
    
    // signature
    public var keyType: String = ""
    public var signKey: DIDURL?
    public var signature: String = ""

    public enum Operation: Int, CustomStringConvertible {
        case CREATE = 0
        case UPDATE = 1
        case DEACTIVATE

        public var description: String {
            let desc: String
            switch self.rawValue {
            case 0:
                desc = "create"
            case 1:
                desc = "update"
            default:
                desc = "deactivate"
            }
            return desc;
        }
    }
    
    private init(_ op: Operation) {
        specification = IDChainRequest.CURRENT_SPECIFICATION
        operation = op
    }
    
    public class func create(_ doc: DIDDocument,
                         _ signKey: DIDURL,
                       _ storepass: String) throws -> IDChainRequest {

        let request = IDChainRequest(Operation.CREATE)
        try request.setPayload(doc)
        try request.seal(signKey, storepass)
        return request
    }
    
    public class func update(_ doc: DIDDocument,
                      previousTxid: String? = nil,
                         _ signKey: DIDURL,
                       _ storepass: String) throws -> IDChainRequest {

        let request = IDChainRequest(Operation.UPDATE)
        request.previousTxid = previousTxid != nil ? previousTxid! : ""
        request.setPayload(doc)
        try request.seal(signKey, storepass)
        return request
    }
    
    public class func deactivate(_ doc: DIDDocument,
                             _ signKey: DIDURL,
                           _ storepass: String) throws -> IDChainRequest {

        let request = IDChainRequest(Operation.DEACTIVATE)
        request.setPayload(doc)
        try request.seal(signKey, storepass)
        return request
    }
    
    public class func deactivate(_ target: DID,
                          _ targetSignKey: DIDURL,
                                    _ doc: DIDDocument,
                                _ signKey: DIDURL,
                              _ storepass: String) throws -> IDChainRequest {

        let request = IDChainRequest(Operation.DEACTIVATE)
        request.setPayload(target)
        try request.seal(targetSignKey, doc, signKey, storepass)
        return request
    }
    
    func setPayload(_ did: DID) {
        self.did = did
        self.doc = nil
        self.payload = did.description
    }
    
    func setPayload(_ doc: DIDDocument) {
        self.did = doc.subject
        self.doc = doc
        
        if operation != Operation.DEACTIVATE {
            let json = doc.description(false)
            let c_input = (json.toUnsafePointerUInt8())!
            payload = json + "\0"
            payload = String(cString: payload.toUnsafePointerUInt8()!)
            let c_payload = UnsafeMutablePointer<Int8>.allocate(capacity: 4096)
            print(payload)
            let re = base64_url_encode(c_payload, c_input, payload.count)
            let jsonStr: String = String(cString: c_payload)
            let endIndex = jsonStr.index(jsonStr.startIndex, offsetBy: re)
            payload = String(jsonStr[jsonStr.startIndex..<endIndex])
        }
        else {
            payload = doc.subject!.description
        }
    }
    
    func setPayload(_ payload: String) {
        if (operation != Operation.DEACTIVATE) {
            let buffer: UnsafeMutablePointer<UInt8> = UnsafeMutablePointer<UInt8>.allocate(capacity: 4096)
            let cp = payload.toUnsafePointerInt8()
            let c = base64_url_decode(buffer, cp)
            var json: String = String(cString: buffer)
            let endIndex = json.index(json.startIndex, offsetBy: c)
            json = String(json[json.startIndex..<endIndex])
            
            doc = try? DIDDocument.fromJson(json)
            did = doc?.subject
        } else {
            did = try? DID(payload)
            doc = nil
        }
        self.payload = payload
    }
    
    func setProof(_ keyType: String, _ signKey: DIDURL, _ signature: String) {
        self.keyType = keyType
        self.signKey = signKey
        self.signature = signature
    }
    
    func seal(_ signKey: DIDURL, _ storepass: String) throws {
        var inputs: [CVarArg] = []
        if specification.count > 0 {
            inputs.append(specification)
            inputs.append(specification.count)
        }
        if operation.description.count > 0 {
            inputs.append(operation.description)
            inputs.append(operation.description.count)
        }
        if previousTxid.count > 0 {
            inputs.append(previousTxid)
            inputs.append(previousTxid.count)
        }
        if payload.count > 0 {
            inputs.append(payload)
            inputs.append(payload.count)
        }
        let count = inputs.count / 2
        self.signature = (try doc?.sign(signKey, storepass, count, inputs))!
        self.signKey = signKey
        self.keyType = DEFAULT_PUBLICKEY_TYPE
    }
    
    func seal(_ targetSignKey: DIDURL, _ doc: DIDDocument, _ signKey: DIDURL, _ storepass: String) throws {
        var inputs: [CVarArg] = []
        let prevtxid = operation == Operation.UPDATE ? previousTxid : ""
        if specification.count > 0 {
            inputs.append(specification)
            inputs.append(specification.count)
        }
        if operation.description.count > 0 {
            inputs.append(operation.description)
            inputs.append(operation.description.count)
        }
        if prevtxid.count > 0 {
            inputs.append(prevtxid)
            inputs.append(prevtxid.count)
        }
        if payload.count > 0 {
            inputs.append(payload)
            inputs.append(payload.count)
        }
        let count = inputs.count / 2
        self.signature = (try doc.sign(signKey, storepass, count, inputs))
        self.signKey = targetSignKey
        self.keyType = DEFAULT_PUBLICKEY_TYPE
    }
    
    public func isValid() throws -> Bool {
        var doc: DIDDocument
        if (operation != Operation.DEACTIVATE) {
            doc = self.doc!
            if (try !doc.isAuthenticationKey(signKey!)){
                return false
            }
        } else {
            doc = try did!.resolve()!
            if (try !doc.isAuthenticationKey(signKey!) && !doc.isAuthorizationKey(signKey!)){
                return false
            }
        }
        var inputs: [CVarArg] = []
        if specification.count > 0 {
            inputs.append(specification)
            inputs.append(specification.count)
        }
        if operation.description.count > 0 {
            inputs.append(operation.description)
            inputs.append(operation.description.count)
        }
        if previousTxid.count > 0 {
            inputs.append(previousTxid)
            inputs.append(previousTxid.count)
        }
        if payload.count > 0 {
            inputs.append(payload)
            inputs.append(payload.count)
        }
        let count = inputs.count / 2
        
        return try doc.verify(signKey!, signature, count, inputs)
    }
    
    public func toJson(_ normalized: Bool) -> String {
        
        var json: OrderedDictionary<String, Any> = OrderedDictionary()
        // header
        var dic: OrderedDictionary<String, Any> = OrderedDictionary()
        dic[IDChainRequest.SPECIFICATION] = specification
        dic[IDChainRequest.OPERATION] = operation.description
        if (operation == Operation.UPDATE) {
            dic[IDChainRequest.PREVIOUS_TXID] = previousTxid
        }
        json[IDChainRequest.HEADER] = dic

        // playload
        json[IDChainRequest.PAYLOAD] = payload
        
        // signature
        var keyId: String
        dic.removeAll(keepCapacity: 0)
        
        if normalized {
            dic[IDChainRequest.KEY_TYPE] = keyType
            keyId = signKey!.description
        }
        else {
            keyId = "#" + signKey!.fragment
        }
        dic[IDChainRequest.VERIFICATION_METHOD] = keyId
        dic[IDChainRequest.SIGNATURE] = signature
        json[PROOF] = dic
        
        let jsonString: String = JsonHelper.creatJsonString(dic: json)
        return jsonString
    }
    
    public class func fromJson(_ json: OrderedDictionary<String, Any>) throws -> IDChainRequest {
        let header = json[HEADER] as! OrderedDictionary<String, Any>
        let spec: String = try JsonHelper.getString(header, SPECIFICATION, false, nil, SPECIFICATION)
        guard (spec == CURRENT_SPECIFICATION) else {
            throw DIDError.failue("Unknown DID specifiction.")
        }
        var opstr: String = try JsonHelper.getString(header, OPERATION, false, nil, OPERATION)
        opstr = opstr.uppercased()
        var op: Operation = .CREATE
        switch opstr {
        case "CREATE": do {
            op = .CREATE
            }
        case "UPDATE": do {
            op = .UPDATE
            }
        case "DEACTIVATE": do {
            op = .DEACTIVATE
            }
        default: break
            
        }
        let request: IDChainRequest = IDChainRequest(op)
        if (op == Operation.UPDATE) {
            let txid = try JsonHelper.getString(header, PREVIOUS_TXID, false, nil, PREVIOUS_TXID)
            request.previousTxid = txid
        }
        let payload: String = try JsonHelper.getString(json, PAYLOAD, false, nil, PAYLOAD)
        request.setPayload(payload)
        
        let proof = json[PROOF] as! OrderedDictionary<String, Any>
        let keyType = try JsonHelper.getString(proof, KEY_TYPE, true,
                                               DEFAULT_PUBLICKEY_TYPE, KEY_TYPE)
        guard (keyType == DEFAULT_PUBLICKEY_TYPE) else {
            throw DIDError.didResolveError(_desc: "Unknown signature key type.") 
        }
        let signKey = try JsonHelper.getDidUrl(proof, VERIFICATION_METHOD, request.did,
                                               VERIFICATION_METHOD)
        let sig = try JsonHelper.getString(proof, SIGNATURE, false,
                                           nil, SIGNATURE)
        request.setProof(keyType, signKey!, sig)
        
        return request
    }
    
    public class func fromJson(_ json: String) throws -> IDChainRequest {
        let string = JsonHelper.preHandleString(json)
        let dic: OrderedDictionary = JsonHelper.handleString(string) as! OrderedDictionary<String, Any>
        return try fromJson(dic)
    }
}

File: DID.Java.SDK/src/main/java/org/elastos/did/backend/IDChainRequest.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did.backend;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;

import org.elastos.did.Constants;
import org.elastos.did.DID;
import org.elastos.did.DIDDocument;
import org.elastos.did.DIDURL;
import org.elastos.did.crypto.Base64;
import org.elastos.did.exception.DIDBackendException;
import org.elastos.did.exception.DIDException;
import org.elastos.did.exception.DIDStoreException;
import org.elastos.did.exception.DIDTransactionException;
import org.elastos.did.exception.InvalidKeyException;
import org.elastos.did.util.JsonHelper;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public class IDChainRequest {
	public static final String CURRENT_SPECIFICATION = "elastos/did/1.0";

	private static final String HEADER = "header";
	private static final String SPECIFICATION = "specification";
	private static final String OPERATION = "operation";
	private static final String PREVIOUS_TXID = "previousTxid";
	private static final String PAYLOAD = "payload";
	private static final String PROOF = "proof";
	private static final String KEY_TYPE = "type";
	private static final String VERIFICATION_METHOD = "verificationMethod";
	private static final String SIGNATURE = "signature";

	private static final String DEFAULT_PUBLICKEY_TYPE = Constants.DEFAULT_PUBLICKEY_TYPE;

	public enum Operation {
		CREATE, UPDATE, DEACTIVATE;

		@Override
		public String toString() {
			return name().toLowerCase();
		}
	}

	// header
	private String specification;
	private Operation operation;
	private String previousTxid;

	// payload
	private DID did;
	private DIDDocument doc;
	private String payload;

	// signature
	private String keyType;
	private DIDURL signKey;
	private String signature;

	private IDChainRequest(Operation op) {
		this.specification = CURRENT_SPECIFICATION;
		this.operation = op;
	}

	public static IDChainRequest create(DIDDocument doc, DIDURL signKey,
			String storepass) throws DIDStoreException, InvalidKeyException {
		IDChainRequest request = new IDChainRequest(Operation.CREATE);
		request.setPayload(doc);
		request.seal(signKey, storepass);

		return request;
	}

	public static IDChainRequest update(DIDDocument doc, String previousTxid,
			DIDURL signKey, String storepass)
			throws DIDStoreException, InvalidKeyException {
		IDChainRequest request = new IDChainRequest(Operation.UPDATE);
		request.setPreviousTxid(previousTxid);
		request.setPayload(doc);
		request.seal(signKey, storepass);

		return request;
	}

	public static IDChainRequest deactivate(DIDDocument doc, DIDURL signKey,
			String storepass) throws DIDStoreException, InvalidKeyException {
		IDChainRequest request = new IDChainRequest(Operation.DEACTIVATE);
		request.setPayload(doc);
		request.seal(signKey, storepass);

		return request;
	}

	public static IDChainRequest deactivate(DID target, DIDURL targetSignKey,
			DIDDocument doc, DIDURL signKey, String storepass)
			throws DIDStoreException, InvalidKeyException {
		IDChainRequest request = new IDChainRequest(Operation.DEACTIVATE);
		request.setPayload(target);
		request.seal(targetSignKey, doc, signKey, storepass);

		return request;
	}

	public Operation getOperation() {
		return operation;
	}

	public String getPreviousTxid() {
		return previousTxid;
	}

	private void setPreviousTxid(String previousTxid) {
		this.previousTxid = previousTxid != null ? previousTxid : "";
	}

	public String getPayload() {
		return payload;
	}

	public DID getDid() {
		return did;
	}

	public DIDDocument getDocument() {
		return doc;
	}

	private void setPayload(DID did) {
		this.did = did;
		this.doc = null;
		this.payload = did.toString();
	}

	private void setPayload(DIDDocument doc) {
		this.did = doc.getSubject();
		this.doc = doc;

		if (operation != Operation.DEACTIVATE) {
			String json = doc.toString(false);

			this.payload = Base64.encodeToString(json.getBytes(),
					Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP);
		} else {
			this.payload = doc.getSubject().toString();
		}
	}

	private void setPayload(String payload) throws DIDTransactionException {
		try {
			if (operation != Operation.DEACTIVATE) {
				String json = new String(Base64.decode(payload,
						Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP));

				doc = DIDDocument.fromJson(json);
				did = doc.getSubject();
			} else {
				did = new DID(payload);
				doc = null;
			}
		} catch (DIDException e) {
			throw new DIDTransactionException("Parse ID operation payload error.", e);
		}

		this.payload = payload;
	}

	private void setProof(String keyType, DIDURL signKey, String signature) {
		this.keyType = keyType;
		this.signKey = signKey;
		this.signature = signature;
	}

	private void seal(DIDURL signKey, String storepass)
			throws DIDStoreException, InvalidKeyException {
		if (!doc.isAuthenticationKey(signKey))
			throw new InvalidKeyException("Not an authentication key.");

		String prevtxid = operation == Operation.UPDATE ? previousTxid : "";

		byte[][] inputs = new byte[][] {
			specification.getBytes(),
			operation.toString().getBytes(),
			prevtxid.getBytes(),
			payload.getBytes()
		};

		this.signature = doc.sign(signKey, storepass, inputs);
		this.signKey = signKey;
		this.keyType = DEFAULT_PUBLICKEY_TYPE;
	}

	private void seal(DIDURL targetSignKey, DIDDocument doc,
			DIDURL signKey, String storepass)
			throws DIDStoreException, InvalidKeyException {
		if (!doc.isAuthenticationKey(signKey))
			throw new InvalidKeyException("Not an authentication key.");

		String prevtxid = operation == Operation.UPDATE ? previousTxid : "";

		byte[][] inputs = new byte[][] {
			specification.getBytes(),
			operation.toString().getBytes(),
			prevtxid.getBytes(),
			payload.getBytes()
		};

		this.signature = doc.sign(signKey, storepass, inputs);
		this.signKey = targetSignKey;
		this.keyType = DEFAULT_PUBLICKEY_TYPE;
	}

	public boolean isValid() throws DIDTransactionException {
		DIDDocument doc = null;
		if (operation != Operation.DEACTIVATE) {
			doc = this.doc;
			if (!doc.isAuthenticationKey(signKey))
				return false;
		} else {
			try {
				doc = did.resolve();
				if (!doc.isAuthenticationKey(signKey) &&
						!doc.isAuthorizationKey(signKey))
					return false;
			} catch (DIDBackendException e) {
				new DIDTransactionException(e);
			}
		}

		String prevtxid = operation == Operation.UPDATE ? previousTxid : "";

		byte[][] inputs = new byte[][] {
			specification.getBytes(),
			operation.toString().getBytes(),
			prevtxid.getBytes(),
			payload.getBytes()
		};

		return doc.verify(signKey, signature, inputs);
	}

	public void toJson(JsonGenerator generator, boolean normalized) throws IOException {
		generator.writeStartObject();

		// header
		generator.writeFieldName(HEADER);
		generator.writeStartObject();

		generator.writeFieldName(SPECIFICATION);
		generator.writeString(specification);

		generator.writeFieldName(OPERATION);
		generator.writeString(operation.toString());

		if (operation == Operation.UPDATE) {
			generator.writeFieldName(PREVIOUS_TXID);
			generator.writeString(previousTxid);
		}

		generator.writeEndObject();

		// payload
		generator.writeFieldName(PAYLOAD);
		generator.writeString(payload);

		// signature
		generator.writeFieldName(PROOF);
		generator.writeStartObject();

		String keyId;

		if (normalized) {
			generator.writeFieldName(KEY_TYPE);
			generator.writeString(keyType);

			keyId = signKey.toString();
		} else {
			keyId = "#" + signKey.getFragment();
		}

		generator.writeFieldName(VERIFICATION_METHOD);
		generator.writeString(keyId);


		generator.writeFieldName(SIGNATURE);
		generator.writeString(signature);

		generator.writeEndObject();
		generator.writeEndObject();
	}

	public void toJson(Writer out, boolean normalized) throws IOException {
		JsonFactory factory = new JsonFactory();
		JsonGenerator generator = factory.createGenerator(out);
		toJson(generator, normalized);
		generator.close();
	}

	public String toJson(boolean normalized) {
		Writer out = new StringWriter(2048);

		try {
			toJson(out, normalized);
		} catch (IOException ignore) {
		}

		return out.toString();
	}

	public static IDChainRequest fromJson(JsonNode node)
			throws DIDTransactionException {
		Class<DIDTransactionException> clazz = DIDTransactionException.class;

		JsonNode header = node.get(HEADER);
		if (header == null)
			throw new DIDTransactionException("Missing header.");

		String spec = JsonHelper.getString(header, SPECIFICATION, false,
				null, SPECIFICATION, clazz);
		if (!spec.equals(CURRENT_SPECIFICATION))
			throw new DIDTransactionException("Unknown ID request specification.");

		String opstr = JsonHelper.getString(header, OPERATION, false,
				null, OPERATION, clazz);
		Operation op = Operation.valueOf(opstr.toUpperCase());

		IDChainRequest request = new IDChainRequest(op);

		if (op == Operation.UPDATE) {
			String txid = JsonHelper.getString(header, PREVIOUS_TXID, false,
					null, PREVIOUS_TXID, clazz);
			request.setPreviousTxid(txid);
		}

		String payload = JsonHelper.getString(node, PAYLOAD, false,
				null, PAYLOAD, clazz);
		request.setPayload(payload);

		JsonNode proof = node.get(PROOF);
		if (proof == null)
			throw new DIDTransactionException("Missing proof.");

		String keyType = JsonHelper.getString(proof, KEY_TYPE, true,
				DEFAULT_PUBLICKEY_TYPE, KEY_TYPE, clazz);
		if (!keyType.equals(DEFAULT_PUBLICKEY_TYPE))
			throw new DIDTransactionException("Unknown signature key type.");

		DIDURL signKey = JsonHelper.getDidUrl(proof, VERIFICATION_METHOD,
				request.getDid(), VERIFICATION_METHOD, clazz);
		//if (doc.getAuthenticationKey(signKey) == null)
		//	throw new DIDResolveException("Unknown signature key.");

		String sig = JsonHelper.getString(proof, SIGNATURE, false,
				null, SIGNATURE, clazz);

		request.setProof(keyType, signKey, sig);
		return request;
	}

	public static IDChainRequest fromJson(String json)
			throws DIDTransactionException {
		if (json == null || json.isEmpty())
			throw new IllegalArgumentException();

		ObjectMapper mapper = new ObjectMapper();
		try {
			JsonNode node = mapper.readTree(json);
			return fromJson(node);
		} catch (IOException e) {
			throw new DIDTransactionException("Parse transaction payload error.", e);
		}
	}

}
